geode-cq/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/DurableClientCQClusterRestartDUnitTest.java

```diff
@@ -0,0 +1,315 @@
+  }
+
+  private String getDiskStoreName() {
+    return getClass().getSimpleName() + VM.getVMId();
+  }
+
+  private void callFunctions() throws Exception {
+    InternalRegion region = (InternalRegion) clientCacheRule.getClientCache().getRegion(regionName);
+    Pool pool = region.getServerProxy().getPool();
+
+    List<Future<Void>> futures = new ArrayList<>();
+    for (int i = 0; i < numOfInvocations; i++) {
+      futures.add(executorServiceRule.submit(() -> invokeFunction(pool)));
+    }
+    for (Future<Void> future : futures) {
+      future.get(getTimeout().toMillis(), MILLISECONDS);
+    }
+  }
+
+  private void invokeFunction(Pool pool) {
+    @SuppressWarnings("unchecked")
+    Execution execution = FunctionService.onServer(pool).setArguments(regionName);
+    ResultCollector resultCollector = execution.execute(new TestFunction());
+    resultCollector.getResult();
+  }
+
+  private void setClientRegion() throws Exception {
+    ClientCache clientCache = clientCacheRule.getClientCache();
+    if (clientCache == null) {
+      Properties config = new Properties();
+      config.setProperty(DURABLE_CLIENT_ID, durableClientId);
+      clientCacheRule.createClientCache(config);
+    }
+
+    Region region = clientCacheRule.getClientCache().getRegion(regionName);
+    if (region != null) {
+      return;
+    }
+    Pool pool = PoolManager.createFactory()
+        .addLocator(hostName, locatorPort)
+        .setSubscriptionEnabled(true)
+        .create(uniqueName);
+
+    ClientRegionFactory crf =
+        clientCacheRule.getClientCache().createClientRegionFactory(ClientRegionShortcut.LOCAL);
+    crf.setPoolName(pool.getName());
+    crf.create(regionName);
+
+    registerCQ();
+  }
+
+  private void registerCQ() throws Exception {
+    ClientCache clientCache = clientCacheRule.getClientCache();
+
+    QueryService queryService = clientCache.getQueryService();
+    CqAttributesFactory cqaf = new CqAttributesFactory();
+    cqaf.addCqListener(new TestCqListener());
+    CqAttributes cqAttributes = cqaf.create();
+
+    queryService.newCq(cqName, "Select * from " + SEPARATOR + regionName,
+        cqAttributes, true).executeWithInitialResults();
+  }
+
+  private void verifyCQListenerInvocations(int expected) {
+    QueryService cqService = clientCacheRule.getClientCache().getQueryService();
+    await().untilAsserted(() -> {
+      CqListener cqListener = cqService.getCq(cqName).getCqAttributes().getCqListener();
+      assertThat(((TestCqListener) cqListener).numEvents.get()).isEqualTo(expected);
+    });
+  }
+
+  public static class TestFunction implements Function, DataSerializable {
+    private final Random random = new Random();
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void execute(FunctionContext context) {
+      CqService cqService = ((InternalCache) context.getCache()).getCqService();
+      waitUntilCqRegistered(cqService);
+
+      String regionName = context.getArguments().toString();
```

Comment:
Warnings here can be fixed by having `TestFunction` implement `Function<String>`, using `FunctionContext<String>` on line 282, and then removing the now-redundant call to `toString()` to get the region name.